src/transformers/models/hunyuan_v1_moe/modular_hunyuan_v1_moe.py

```diff
@@ -0,0 +1,358 @@
+        drop_idx = drop_mask.nonzero().view(-1)
+        drop_num = drop_mask.sum().to(torch.int64)
+        indices1_s_after_drop.scatter_(0, drop_idx, not_full_experts_ids[:drop_num])
+        nodrop_mask1 = F.one_hot(indices1_s_after_drop, num_classes=num_experts)
+        mask1 = nodrop_mask1
+
+    # Compute locations in capacity buffer
+    locations1 = torch.cumsum(mask1, dim=0) - 1
+
+    # Store the capacity location for each token
+    locations1_s = torch.sum(locations1 * mask1, dim=1)
+
+    # Normalize gate probabilities
+    mask1_float = mask1.float()
+    gates = gates * mask1_float
+
+    locations1_sc = F.one_hot(locations1_s, num_classes=capacity).float()  # one hot to float
+    combine_weights = torch.einsum("se,sc->sec", gates, locations1_sc)
+
+    dispatch_mask = combine_weights.bool()
+
+    exp_counts_capacity = torch.sum(mask1_bk)
+    exp_capacity_rate = exp_counts_capacity / (logits.shape[0])
+    return [l_aux, exp_capacity_rate], combine_weights, dispatch_mask, exp_counts
+
+
+class HunYuanTopKGate(nn.Module):
+    def __init__(self, config: HunYuanMoEV1Config, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.moe_topk = config.moe_topk if isinstance(config.moe_topk, int) else config.moe_topk[layer_idx]
+        self.drop_tokens = config.moe_drop_tokens
+        self.random_routing_dropped_token = config.moe_random_routing_dropped_token
+        num_experts = config.num_experts if isinstance(config.num_experts, int) else config.num_experts[layer_idx]
+        self.wg = nn.Linear(config.hidden_size, num_experts, bias=False, dtype=torch.float32)
+
+    def forward(self, hidden_states):
+        bsz, seq_len, hidden_size = hidden_states.shape
+        hidden_states = hidden_states.reshape(-1, hidden_size)
+        if self.wg.weight.dtype == torch.float32:
+            hidden_states = hidden_states.float()
+        logits = self.wg(hidden_states)
+        if self.moe_topk == 1:
+            gate_output = top1gating(logits, random_routing_dropped_token=self.random_routing_dropped_token)
+        else:
+            gate_output = topkgating(logits, self.moe_topk)
+
+        return gate_output
+
+
+class HunYuanMoE(nn.Module):
+    def __init__(self, config: HunYuanMoEV1Config, layer_idx: Optional[int] = None):
+        super().__init__()
+        self.config = config
+        self.layer_idx = layer_idx
+        self.moe_topk = config.moe_topk
+        self.num_experts = config.num_experts if isinstance(config.num_experts, int) else config.num_experts[layer_idx]
+        if config.use_mixed_mlp_moe:
+            self.shared_mlp = HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=True)
+        self.gate = HunYuanTopKGate(config, layer_idx=layer_idx)
+        self.experts = nn.ModuleList(
+            [HunYuanMoEV1MLP(config, layer_idx=layer_idx, is_shared_mlp=False) for _ in range(self.num_experts)]
```

Comment:
modulelist is worst than having 3d weights will help rather than modulist ! Don't you want to fix this to look more like gpt oss / llama4 ? will help with vllm as well